var/spack/repos/builtin/packages/python/package.py

```diff
@@ -591,11 +591,21 @@ def setup_dependent_environment(self, spack_env, run_env, dependent_spec):
         pythonpath = ':'.join(python_paths)
         spack_env.set('PYTHONPATH', pythonpath)
 
-        # For run time environment set only the path for
-        # dependent_spec and prepend it to PYTHONPATH
+        # For run time environment set path for all dependent_spec
+        # recursively and prepend it to PYTHONPATH
+        python_paths = []
         if dependent_spec.package.extends(self.spec):
-            run_env.prepend_path('PYTHONPATH', join_path(
-                dependent_spec.prefix, self.site_packages_dir))
+            path = join_path(dependent_spec.prefix, self.site_packages_dir)
+            python_paths.append(path)
+        for em in run_env.env_modifications:
+            if em.name == 'PYTHONPATH':
+                python_paths.append(em.value)
+        for d in dependent_spec.traverse(deptype=('run')):
+            if d.package.extends(self.spec):
+                path = join_path(d.prefix, self.site_packages_dir)
+                python_paths.append(path)
+        pythonpath = ':'.join(python_paths)
+        run_env.set('PYTHONPATH', pythonpath)
```

Comment:
Maybe I'm missing something, but shouldnt we use prepend_path instead?

Reply:
I got confused with this as well. I followed the logic of spack_env and it works because `setup_dependent_environment` is called for every package that we are installing. And as we traverse **all dependencies** for that package and set PYTHONPATH. And hence we don't need to prepend but just set. (I think if we prepend then it will grow more than necessary, for example, **every time** each package will append it's dependencies path).

Reply:
What happens though if you load multiple modules this way, with non-perfectly overlapping dependencies?

Reply:
Ouch! to handle this I think we should change this to `append/prepend`.

also, see upstream comments in spack/spack/pull/9527 and spack/spack/issues/8639. But if this PR is convenient in our use cases, I am in favour of going ahead with this logic.